atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/SchemaMutationLock.java

```diff
@@ -68,18 +66,102 @@ public SchemaMutationLock(
+
+        void startBeating() {
+            keepBeating = true;
+            heartbeatThread.start();
+        }
+
+        void stopBeating() {
+            keepBeating = false;
+            try {
+                // make sure beating's stopped
+                heartbeatThread.join();
+            } catch (InterruptedException e) {
+                LOGGER.debug("Interrupted while trying to stop the heartbeat heartbeatThread.");
+            }
+        }
+
+        void heartbeat() {
+            while (keepBeating) {
+                try {
+                    clientPool.runWithRetry((FunctionCheckedException<Cassandra.Client, Void, Exception>) client -> {
+                        Column ourUpdate = lockColumnWithStrValue(String.format(
+                                CassandraConstants.GLOBAL_DDL_LOCK_FORMAT, lockId, heartbeatCount + 1));
+
+                        List<Column> expected = ImmutableList.of(lockColumnWithStrValue(String.format(
+                                CassandraConstants.GLOBAL_DDL_LOCK_FORMAT, lockId, heartbeatCount)));
+
+                        CASResult casResult = writeDdlLockWithCas(client, expected, ourUpdate);
+
+                        if (casResult.isSuccess()) {
+                            heartbeatCount++;
+                        } else {
+                            handleForcedLockClear(casResult, lockId, heartbeatCount);
+                        }
+                        return null;
+                    });
+                    Thread.sleep(configManager.getConfig().heartbeatTimePeriodMillis());
+                } catch (Exception e) {
+                    throw Throwables.throwUncheckedException(e);
+                }
+            }
+        }
+    }
 
+    void runWithLock(Action action) {
+        if (!supportsCas) {
+            runWithLockWithoutCas(action);
+            return;
+        }
+
+        long lockId = waitForSchemaMutationLock();
+        heartbeatThread = new HeartbeatThread(lockId);
+        heartbeatThread.startBeating();
         try {
             action.execute();
         } catch (Exception e) {
             throw Throwables.throwUncheckedException(e);
         } finally {
+            heartbeatThread.stopBeating();
             schemaMutationUnlock(lockId);
         }
     }
 
+    private void runWithLockWithoutCas(Action action) {
+        if (configManager.getConfig().servers().size() > 1) {
+            throw new UnsupportedOperationException("Running a clustered service with a version of Cassandra"
+                    + " that does not support check and set is not allowed. Either upgrade Cassandra or run"
+                    + " a single node service");
+        }
+        LOGGER.info("Because your version of Cassandra does not support check and set,"
+                + " we will use a java level lock to synchronise schema mutations.");
+        try {
+            waitForSchemaMutationLockWithoutCas();
+            action.execute();
+        } catch (Exception e) {
```

Comment:
I.e. We probably want to test cancellation and determine if we should be resetting the current thread's interrupt flag.